### DISCUSSION TITLE
[V3] Initial event does not update component attribute

### DISCUSSION THREAD

**Discussion by Drenth1** [livewire/livewire#6516] [7. Bugs (archived)] [answered]
### Livewire version

v3.0.0

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

I have two Livewire components that I want to interact with eachother. A panel with filters and a table with users. I removed all irrelevant code from the snippets below. Both components have a class and return a blade view. This is the filter component:

```html
<!-- livewire/filters/user-filter.blade.php -->

<input type="text" name="search" placeholder="Search users" wire:model.lazy="searchTerm"/>
```

```php
// ... App\Livewire\Filters\UserFilter.php

class UserFilter extends Component
{
    /**
     * The currently active filter.
     *
     * @var string
     */
    public string $searchTerm = '';

    /**
     * Render the component view.
     *
     * @return \Illuminate\View\View
     */
    public function render() : View
    {
        return view('livewire.filters.user-filter');
    }

    /**
     * Actions to run when updating the searchTerm attribute.
     *
     * @return void
     */
    public function updatingSearchTerm() : void
    {
        $this->dispatch('filters.user.searchTerm.update', searchTerm: $this->searchTerm);
    }
}
```

This is (the essential part of) the table component:

```blade
<!-- livewire/tables/user-table.blade.php -->

@forelse($users as $user)
    <tr>
        <td>{{ $user->name }}</td>
        // ... other attributes
    </tr>
@endforelse
```

```php
// ... App\Livewire\Tables\UserTable.php

class UserTable extends Component
{
    use WithPagination;

    public ?string $searchTerm = null;

    /**
     * Render the component view.
     *
     * @return \Illuminate\View\View
     */
    public function render() : View
    {
        $users = empty($this->searchTerm)
            ? User::paginate(10)
            : User::whereLike(['name', 'email'], $this->searchTerm)->paginate(10);

        return view('livewire.tables.user-table', [
            'users' => $users
        ]);
    }

    /**
     * Filter the users by a search term.
     *
     * @param string $searchTerm
     * @return void
     */
    #[On('filters.user.searchTerm.update')]
    public function filterUsers(string $searchTerm) : void
    {
        $this->searchTerm = $searchTerm;
    }
}
```

When typing something in the search input and pressing enter I _sometimes_ get the following error in the browser console:

> Uncaught TypeError: currentNode is null

And a bunch of these warnings:

> unreachable code after return statement

### Additional information

Some things to note:

1. When changing `wire:model.lazy` to `wire:model` the event does not fire at all

2. Editing the update function reveals something odd, the variable seems to be one request behind or something:

- If I enter `Hello` in the search bar and press enter it shows an empty string
- If I then enter `Hello2` and press enter it shows `Hello`
- If I then enter `Hello3` and press enter it shows `Hello2`

```php
public function updatingSearchTerm() : void
{
    dump($this->searchTerm);
    $this->dispatch('filters.user.searchTerm.update', searchTerm: $this->searchTerm);
}
```

When using proper terms that should yield results, the table updates, albeit after I send another request with a different search term. It uses the search term of one request prior.

3. `WhereLike` is a custom builder macro.
4. The initial render of the component on page load works fine.
5. This project uses only Livewire, not AlpineJS
6. Dispatching the event to the component directly does not change anything

7. The event looks like it fires twice:

This is the first occurrence with the proper search term:
![Screenshot_67](https://github.com/livewire/livewire/assets/89132520/19921ee3-5643-454c-a1b5-2f2c89d2ae61)

This is the second occurrence with an empty search term:
 
![Screenshot_68](https://github.com/livewire/livewire/assets/89132520/bea75365-3644-4ef3-88ab-6246aeeccf81)

If any additional information is needed, please let me know.

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by Drenth1** [comment]
Changing the `@forelse` to a `@foreach` fixes the `Uncaught TypeError: currentNode is null` error. However, the issue with the variable updating one request later remains.

**Reply by gofish543** [reply]
Oh my god I'm glad I'm not the only one with this one request deferred issue!

**Comment by Drenth1** [comment, marked as answer]
@gofish543 The thing that fixed the initial request bugging out was changing this:

```php
public function updatingSearchTerm() : void
{
    $this->dispatch('filters.user.searchTerm.update', searchTerm: $this->searchTerm);
}
```

to the version as described in [the v3 documentation](https://livewire.laravel.com/docs/lifecycle-hooks#update):

```php
public function updating(string $property, mixed $value) : void
{
    if ($property === 'searchTerm')
    {
        $this->searchTerm = $value;
        $this->dispatch('filters.user.searchTerm.update', searchTerm: $this->searchTerm);
    }
}
```
